app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala

```diff
@@ -24,4 +24,5 @@ import uk.gov.hmrc.model.NonEmptyString
 case class PersonalDetails(firstName: NonEmptyString,
```

Comment:
This domain object is very loose now. It is easy to create an instance of `PersonalDetails ` case class by providing both nino and postcode.
I propose having an ADT (algebraic data type) structure with two case classes `PersonalDetailsWithPostcode` and `PersonalDetailsWithNino` (or better names which can convey the right meaning) having a common base trait `PersonalDetails`. This will not let anybody create invalid domain objects.

Reply:
Few more thing for using separate types for both cases would be that we will have less 'if' conditions spread in different places, more compile-time safety and then fewer unit tests. Effectively, because the requirement lists two different excluding cases which have to be dealt with in slightly different ways, it somehow seems to me more natural to model it with two types.

Reply:
We have the requirement to update the validation resource with the Nino, so there will be cases where we have nino and postcode.

Reply:
@steve-edmans - The Validation resource itself will contain only the NINO. The input request body will either contain the postcode or the NINO. I don't see how we can have both of them together.